### ISSUE TITLE
Exploration Inconsistency when Corresponding Table is Deleted

### ISSUE THREAD

**Issue by Aritra8438** [centerofci/mathesar#2747]
## Description
When we create an exploration, we need to choose a _From_ table. If that table is deleted, the exploration based on that is not deleted automatically (in frontend) resulting in inconsistency (let's say them **dangling explorations**). if we then open the exploration, there shows an error ```Invalid pk "193" - object does not exist.```

Adding a video file to better understand the situation (error view in 30th second):

https://user-images.githubusercontent.com/64671908/227758333-1361466c-9ced-4aa8-bd34-36f2b2d6e6c2.mp4

## Expected behavior
There are some observations: 
- Even after the table is deleted, **we can edit the dangling exploration and assign it to a new table**. But we can't save it after the reassignment (throws error ```Not Found```)
- There needs to be at least one existing table for that, otherwise we need to delete the exploration.
- Once reloaded, all the dangling explorations are removed from the explorations tab (that means, they are deleted from database).

Based on those, we may have two solutions,

1. Delete all those dangling explorations from the explorations tab once the corresponding tables are deleted.
2. Allow assigning dangling explorations to another table and **successfully save them**.

In both cases, the minimum we need is not to show an error ```Invalid pk "193" - object does not exist.``` that might not be interpreted by a user but rather to show them the root cause and alternative solutions.

## To Reproduce
- create a table, say ```table_1```
- create an exploration from ```table_1```, say ```exploration_1```
- delete ```table_1```
- now open ```exploration_1```, it will show an error.
- reload the page, those dangling explorations won't show up again.

## Environment
 - OS: Ubuntu 22.04.1 LTS
 - Browser: Google Chrome
 - Browser Version: 111.0.5563.111 (Official Build) (64-bit)

**Comment by Aritra8438** [comment]
I want to work on this issue as well (if labels are appropriate). Also, please suggest a solution that you think will be better.
The backend deletes the explorations once the corresponding table is deleted, but keeping those dangling explorations and allowing reassignment would lessen the user modification needed.

**Comment by rajatvijay** [comment]
@Aritra8438 I prefer the first option. 

> Delete all those dangling explorations from the explorations tab once the corresponding tables are deleted.

But note this only requires to re-fetch the queries for the schema once a table is deleted. 

> In both cases, the minimum we need is not to show an error Invalid pk "193" - object does not exist. that might not be interpreted by a user but rather to show them the root cause and alternative solutions.

I don't think we need to show any error if we re-fetch the queries. 

In addition to this, I also suggest showing some kind of warning to the user inside the delete table confirmation modal by saying something like: 

> Deleting this table will also delete all of the explorations built on top of it. Are you sure you want to proceed?


cc: @pavish adding this to v0.1.3 since this seems important from the perspective of UX.

**Comment by Aritra8438** [comment]
Sure, @rajatvijay, I will work on that. Refetching the explorations would solve this problem. Thank you.

**Comment by Aritra8438** [comment]
> adding this to v0.1.3 since this seems important from the perspective of UX.

   Hi @rajatvijay, are **backlogs** prioritized over **v0.1.3**?

Also, to fix this bug I need to modify the files which are already modified in [this PR](https://github.com/centerofci/mathesar/pull/2745). So, fixing this bug and making a pull request may generate **merge conflicts**. 
Should I proceed, or wait for the merge approval of the existing PR?

**Comment by kgodey** [comment]
@rajatvijay Please talk to @ghislaineguerin about the expected behavior here. I think the idea behind leaving the dangling explorations was to allow the user to recover from errors without needing to start from scratch.

**Comment by pavish** [comment]
@kgodey I think this might be a valid bug, not an intentional UX.

I remember the behaviour was that if a table is deleted, we would delete the associated exploration as well. However, when a column is deleted, we would show errors to help fix the exploration. Perhaps this is related to metadata changes and we somehow did not catch this.

Even if we did want to have dangling explorations, the error shown here is 'Invalid pk' instead of something like 'Table not found'.

**Comment by kgodey** [comment]
@pavish I agree this is a bug, I just wanted to double check what the intended UX should be (maybe there's something better than deleting the danging explorations?). I don't have time to follow up further on this, so I'll leave it to your and @ghislaineguerin's judgement.

**Comment by ghislaineguerin** [comment]
I remember when we discussed this with @mathemancer, the idea was to leave the exploration, even if the table had been deleted and allow users to point to another table, to replace it and not lose their exploration. Imagine if they had deleted by mistake then recover it, removing the dangling explorations could be very damaging to the user.

**Comment by pavish** [comment]
@ghislaineguerin It's not possible to recover an exploration where the base table is deleted, pointing to a different table (even if it's the same table that is recovered) will not be possible since the columns used are entirely different. We don't have  If we **have** to do that, that requires an entire new user flow to map columns to new table which would be cumbersome with links involved.

**Comment by Aritra8438** [comment]
@pavish, so we are going with **deleting those dangling expolorations**? 
Should I start working on this issue?

**Comment by mathemancer** [comment]
@pavish If we automatically delete explorations whenever underlying dependencies are deleted, it's not possible for users to recover through any means. If we don't automatically delete such explorations, then (even if they have to manually open up another tab and create another exploration based on the now-broken one) they can try to help themselves recover.

That said, I defer to @ghislaineguerin 's judgement either way.

**Comment by ghislaineguerin** [comment]
I maintain the idea that removing explorations without notifying users results in a poor experience, as it doesn't communicate the system status accurately. Users should know their explorations are broken or invalid rather than missing. This is important because tables may be deleted outside of Mathesar.

By leaving the explorations in place, users can identify those that are invalid and delete them manually, as there are no alternative recovery methods.

The ideal would be that they can recover, but since that isn't happening now, then we can at least try to keep the information on how that exploration was built if possible.

**Comment by pavish** [comment]
@mathemancer @ghislaineguerin 

I understand to not deleting an exploration when other dependencies are deleted. The issue here is when a base table is deleted. All the required endpoints to show an error do not work when the base table is deleted, mainly the get endpoint for the exploration, the run shows a different error, and the joinable_tables endpoint (which is fine). This is incontrast to deleting a joined table which will show an error with the missing columns and ability to recover it.

> If we don't automatically delete such explorations, then (even if they have to manually open up another tab and create another exploration based on the now-broken one) they can try to help themselves recover.
> we can at least try to keep the information on how that exploration was built if possible.

So, users are already unable to recover from this case in any means, or understand why it fails.

Having said all of this, if the backend can support recovery or atleast an ability to return the proper error, the frontend can support this. My concern was the viability of supporting this on the backend rather than the UX.

**Comment by pavish** [comment]
I'm restricting this to maintainers, since whatever we decide here would require design work, backend work, and frontend work.

**Comment by pavish** [comment]
I'll like one last confirmation on the backend viability from @mathemancer before unmarking this from draft. When the base table is deleted, can the APIs respond in a similar way as to when a joined table is deleted?

@ghislaineguerin The UX updates we'll need for this particular state may have complications and would need review/confirmation from the backend team:
- A proper error message
- A recovery option if possible
- The base table is no longer present. What happens on the base table selector?
- If the base table selector is empty, what happens on the column selection & transformation panes?

For comparison, here's the difference when base table is deleted vs joined table is deleted:

| Base table is deleted | Joined table is deleted |
|---|---|
| <img width="1436" alt="Screenshot 2023-04-26 at 6 00 52 PM" src="https://user-images.githubusercontent.com/11850603/234575914-b0dd05e8-68a3-4a1a-a6c4-b42e0a2982c0.png">|<img width="1436" alt="Screenshot 2023-04-26 at 6 02 44 PM" src="https://user-images.githubusercontent.com/11850603/234575930-0755ae3e-92b7-4d11-99a8-6410d132e44d.png">  |
| GET endpoint for query fails | GET endpoint returns the query |
| Run endpoint fails with random errors | Run endpoint return proper error |
| Base table select menu on the UI cannot show the selected table, because the information no longer present anywhere | The base table information is shown properly on the UI |
| Query cannot be recovered | Query can be recovered in some manner |
